witness.md

```diff
@@ -0,0 +1,599 @@
+
+    - the pattern to match against;
+
+- result, to the right of the `|=>` sign.
+
+
+### `GUARD`s
+
+Each substitution rule can have zero, one or multiple `GUARD` statements.
+Each `GUARD` statement looks like this:
+
+```
+GUARD <CONDITION>
+```
+
+For a substitution rule to be applicable, the `<CONDITION>` in its `GUARD` statement MUST be true.
+
+If a substitution rule has multiple `GUARD` statements, all of them MUST BE satisfied.
+
+If there are no `GUARD` statements, the substitution rule's applicability is
+only defined by the PATTERN.
+
+Example:
+```
+ GUARD NBITSET(mask) == 2
+|---- GUARD STATEMENT ---|
+
+ Node(n0) Node(n1) BRANCH(mask) |=> 
+ BranchNode{MAKE_VALUES_ARRAY(mask, n0, n1)}
+```
+
+For the example rule to be applicable both facts MUST be true:
+
+1. `mask` contains only 2 bits set to 1 (the rest are set to 0);
+
+2. to the left of `BRANCH` instruction there is at least 2 `Node`s.
+
+Fact (1) comes from the `GUARD` statement.
+
+
+### PATTERN
+
+`[NodeType(boundVar1)... NodeType(boundVarN)] Instruction[(param1... paramN)]`
+
+A pattern MUST contain a single instruction.
+A pattern MAY contain one or more `Node`s to the left of the instruction to
+match.
+An instruction MAY have one or more parameters.
+
+Pattern matching is happening by the types. `Node` type means any node is
+matched, some specific node type will require a specific match.
+
+Pattern can have bound variable names for both matched nodes and instruction
+parameters (if present).
+
+Match:
+
+```
+HASH h0 HashNode{h1} HashNode{h2} BRANCH 0b11
+       |------------------- MATCH -----------|
+
+HASH h0 BranchNode{0: HashNode{h1} 1: HashNode{h2}}
+       |----------- SUBSTITUTED -------------------|
+```
+
+No match (not enough nodes to the left of the instruction):
+
+```
+HASH h0 HASH h1 HashNode{h2} BRANCH 0b11
+```
+
+### Result
+
+`NodeType(HELPER_FUNCTION(arguments))`
+
+The result is a single `Node` statement that replaces the pattern in the
+witness if it matches and the guards are passed.
+
+The result MAY contain helper functions or might have in-line computation.
+The result MUST have a specific node type. No generic `Node` is allowed.
+
+Helper functions or inline computations might use bound variables from the
+pattern. 
+
+Example
+
+```
+                             
+Node(n0) Node(n1) BRANCH(mask) |=>
+BranchNode{MAKE_VALUES_ARRAY(mask, n0, n1)}
+                             ^     ^-- ^--- BOUND NODES
+                             |---- BOUND INSTRUCTION PARAM
+          |------ HELPER CALL ------------|
+|----------------- RESULT ------------------|
+
+```
+
+### Bringing it all together
+
+
+So the full syntax is this:
+
+```
+[GUARD <CONDITION> ...] [ NodeType(bound_variable1)... ] INSTRUCTION[(param1 ...)] |=>
+Node(<HELPER_FUNCTION_OR_COMPUTATION>)
+```
+
+`NodeType` is one of the types of nodes to match. Can also be `Node` to match
+any non-nil node.
+
+Substitution rules MUST be non-ambiguous. Even though, there can be multiple
```

Comment:
Saying "Substitution rules MUST be non-ambiguous" is hopeful. This must be a proven property of your syntax.